eth_abi/registry.py

```diff
@@ -0,0 +1,127 @@
+import functools
+import operator
+
+from .exceptions import (
+    NoEntryFound,
+)
+
+
+class PredicateMapping:
+    """
+    Acts as a mapping from predicate functions to values.  Values are retrieved when their
+    corresponding predicate matches a set of given inputs.  Predicates can also be labeled to
+    facilitate removal from the mapping.
+    """
+    def __init__(self):
+        self._values = {}
+        self._labeled_predicates = {}
+
+    def add(self, predicate, value, label=None):
+        if predicate in self._values:
+            raise ValueError('Predicate already exists')
+
+        if label is not None:
+            if label in self._labeled_predicates:
+                raise ValueError('Predicate with label "{}" already exists'.format(label))
+
+            self._labeled_predicates[label] = predicate
+
+        self._values[predicate] = value
+
+    def find(self, type_str):
+        results = tuple(
+            value for predicate, value in self._values.items()
+            if predicate(type_str)
+        )
+
+        if len(results) == 0:
+            raise ValueError('No matching values found')
+
+        if len(results) > 1:
+            raise ValueError('Multiple matching values found')
+
+        return results[0]
+
+    def remove_by_reference(self, predicate):
+        # Delete the predicate mapping to the previously stored value
+        try:
+            del self._values[predicate]
+        except KeyError:
+            raise KeyError('Predicate not found')
+
+        # Delete any label which refers to this predicate
+        try:
+            label = self._label_for_predicate(predicate)
+        except ValueError:
+            pass
+        else:
+            del self._labeled_predicates[label]
+
+    def _label_for_predicate(self, predicate):
+        # Both keys and values in `_labeled_predicates` are unique since the
+        # `add` method enforces this
+        for key, value in self._labeled_predicates.items():
+            if value is predicate:
+                return key
+
+        raise ValueError('Predicate not referred to by any label')
+
+    def remove_by_label(self, label):
+        try:
+            predicate = self._labeled_predicates[label]
+        except KeyError:
+            raise KeyError('Label not found')
+
+        del self._labeled_predicates[label]
+        del self._values[predicate]
+
+    def remove(self, predicate_or_label):
+        if callable(predicate_or_label):
+            self.remove_by_reference(predicate_or_label)
+        elif isinstance(predicate_or_label, str):
+            self.remove_by_label(predicate_or_label)
+        else:
+            raise TypeError('Key to be removed must be a callable or a string')
```

Comment:
This is another place where including the type that was received in the error message makes debugging this failure path easier later.  Something like `"...: Got {0}".format(type(predicate_or_label))`